### PR TITLE
Upload list activity: integrated overflow menu

### LINKED ISSUE #3463: Improvement of former Expert Mode features
* swipe to refresh for force re-scan
* trash bin icon on uploaded / failed row
* retry button failed (or should this be done via swipe to refresh? )

cc @tobiasKaminsky

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
Fix #3463

- swipe to refresh for force re-scan & retry upload
- trash bin icon to delete all entries
- used butterknife

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
👍 Looking good (except for the raised lint/findbugs scores) ❤️

**Comment by tobiasKaminsky** [comment]
As suggested by lint, I combined text and trash bin icon, now the whole header is clickable and will remove the entries
(I guess the tiny trashbin itself was also a bit tiny)

**Comment by tobiasKaminsky** [comment]
Ah. Hm. Maybe this will confuse people, as they might expect that clicking on the header text will collapse?

So what about this:
- click on trash bin removes entry
- click on group name will collapse?

**Comment by AndyScherzinger** [comment]
hmm, collapse seems like a nice idea/addition, whiel we then need to store this state so it'll be restored when entering the upload screen at a later point again.
We should then maybe also put the trash icon to the right-end of the header line too (like we do it for the icons on the file listing)

**Comment by tobiasKaminsky** [comment]
> hmm, collapse seems like a nice idea/addition, whiel we then need to store this state so it'll be restored when entering the upload screen at a later point again.

Not sure if we need this, but yeah we could add this of course.
Shall I? :-D

**Comment by AndyScherzinger** [comment]
we don't need it but like we both think: it would be a nice addition, so _if_ you have time to take care of it, just go for it 🥇

**Comment by AndyScherzinger** [comment]
I general I am fine with empty categories. In case all categories are empty then we should display the empty view like we do now.

**Comment by tobiasKaminsky** [comment]
> I general I am fine with empty categories. In case all categories are empty then we should display the empty view like we do now.

Seems to be far too easy. There is already a "toggleSectionExpanded" function :+1: 
I'll add it and fix Codacy.

File: src/main/java/com/owncloud/android/ui/adapter/UploadListAdapter.java [hunk 1]
```diff
@@ -188,31 +210,31 @@ public void onBindViewHolder(SectionedViewHolder holder, int section, int relati
...
                         // not really uploading; stop listening progress if view is reused!
-                        if (mProgressListener != null && mProgressListener.isWrapping(itemViewHolder.progressBar)) {
-                            binder.removeDatatransferProgressListener(mProgressListener,
-                                    mProgressListener.getUpload()   // the one that was added
+                        if (progressListener != null && progressListener.isWrapping(itemViewHolder.progressBar)) {
+                            binder.removeDatatransferProgressListener(progressListener,
+                                                                      progressListener.getUpload()   // the one that was added
                             );
-                            mProgressListener = null;
+                            progressListener = null;
```
[the remark this answers is not in the document]
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/adapter/UploadListAdapter.java]: Out of scope

**Comment by jmue** [comment]
To be honest, this interface looks crude. You have trash icons in two different sizes und at two different vertical alignments.

Maybe this is not perfect, but what do you think of having the trash icon in the headerbar und using some select mimic (click on header -> select all items in that group)?

**Comment by AndyScherzinger** [comment]
@jmue thanks for the feedback and input, you are right, this should be optimized.

thus:
@nextcloud/designers can you please have a look at https://github.com/nextcloud/android/pull/3472#issuecomment-455516257 for designer input.

**Comment by tobiasKaminsky** [comment]
Thank you for your input, indeed the UI looks a bit off.

> having the trash icon in the headerbar und using some select mimic (long press on icons / long press on header -> select all items in that group)?

We always try to make the UI as clear as possible, having two different actions on the same element, but without showing the user what is doing what, is in my opinion bad UX.

Hopefully our designer team will find another good way :-)

**Comment by jancborchardt** [comment]
Some input :)

- We shouldn’t use the trash icon here, as it stands for "remove". Having a trash icon next to a file or a list of files looks like it will remove _the file_, not only the entry.
- We should rather use the [icon-close x](https://docs.nextcloud.com/server/latest/developer_manual/design/icons.html).
- We don’t need the trash icon on every single entry I’d say, it’s too much?
- The icons should have action text so it’s understandable, and the action should be on the right of the header row. Tapping the header text itself should not carry out that action, it doesn’t need to be interactive.
- When the sublist is empty like Uploaded (0), there does not need to be any icon.
- On "Current" we don’t need any icon at all? Either it will be successful, or it will fail. Maybe a "Pause/Resume" icon if at all?
- "Failed & pending" should have a [retry/refresh](https://material.io/tools/icons/?search=refresh&icon=refresh&style=baseline) icon.

In summary, wording and icons:
- Current: "⏸ Pause all"
- Failed & pending: "🔁 Retry all"
- Uploaded: "❌ Clear all"

--------------------------

And in the big picture – let me know what you think :)
- The upload view should be more integrated in the normal list, maybe even to the point where the separate section can be removed.
- Basically the "Uploaded" list already exists via "Sort by modified" and "Recent" also. Sure it includes edited files too but it’s fine.
- Errors of any kind should be presently displayed, not only in Uploads.
- For current uploads we already have the notification too, and they should be displayed as "Ghost files" in the location where they are to be placed.

This would integrate it well and do away with the need for a separate section.

**Comment by AndyScherzinger** [comment]
@jancborchardt how would you then do 
- Failed & pending: "❌ clear all"?

Else there is now way to ever get rid of these items (e.g. deleted items that have been scheduled for upload, so no way to ever retry). So for failed/pending there is actually more than one action.

All out "actions" are either icons-only or textonly the later being used within menus. So with this proposale do you want to move to icon+text which might then well rather be outlined buttons with icon+text (aka our secodary action buttons + icons). This is how Google's files app is doing this, see:
screenshot_20190122-172417_files
While we would of course give it a pill-style :)

**Comment by jancborchardt** [comment]
> Else there is now way to ever get rid of these items (e.g. deleted items that have been scheduled for upload, so no way to ever retry). So for failed/pending there is actually more than one action.

If the only thing you can do with them is manually remove them from the list, then they should just be cleared from the list automatically. No need to show files which failed to upload which don’t exist anymore anyway.


> All out "actions" are either icons-only or textonly the later being used within menus. So with this proposale do you want to move to icon+text which might then well rather be outlined buttons with icon+text (aka our secodary action buttons + icons). This is how Google's files app is doing this

Sure, having them be buttons is good too! :) The text is definitely important.

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger do you have found a solution with Jan and can implement this? 
I am a bit lost…

@jancborchardt the big picture is nice, but totally out of scope for this ;-)

**Comment by AndyScherzinger** [comment]
> do you have found a solution with Jan and can implement this?
I am a bit lost…

@tobiasKaminsky assigned it to me and yes, I do have an idea how to solve this, at least to the next increment which might still have issues for clearing failed uploads but would sacrifice this for now to get the change build (probably in the next 2 weeks9 and then further push the discussion since wew likely have several situations for failed uploads where we might want to enable retries _and_ deletion of these entries. We'll see.

Hope that works for you @tobiasKaminsky :)

**Comment by gloschtla** [comment]
I miss the "Delete All" button. I have about thousand of failed uploads in that list.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky can you give it anothere spin? I did all the changes that can be done at this very moment. For anything else I'll need to create new, separate issues since these changes need further features to even work (like a pause for uploads). So these UI changes can't be done at the moment but we shouldn't loose the specification we have here (in the comments).

**Comment by AndyScherzinger** [comment]
@rullzer why is your codacy config/project hijacking this repo / build checks?

**Comment by rullzer** [comment]
Those were added on the request of @tobiasKaminsky. I can kill them off if needed.

**Comment by AndyScherzinger** [comment]
@rullzer what is mean is:
https://app.codacy.com/app/dan0xii/android/pullRequest?prid=2936746
Vs.
https://app.codacy.com/app/rullzer/android/pullRequest?prid=2936746
Vs.
https://app.codacy.com/app/Nextcloud/android/pullRequest?prid=2936746

It should only be the last one while for some reason the status seems to flip through the different codacy configs someone created 0_o

**Comment by rullzer** [comment]
mmm weird. ANyways I removed my project.

**Comment by tobiasKaminsky** [comment]
Fine & works :+1: 
(is the padding on the right big enough?)

**Comment by AndyScherzinger** [comment]
> (is the padding on the right big enough?)

I think so, the icons are aligned to the menu in the toolbar (like we also do for e.g. the file listing)

**Comment by tobiasKaminsky** [comment]
Ha, I knew that I should not doubt on you ;-)

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky so merge then? 😉

**Comment by AndyScherzinger** [comment]
❗️ rebased...

@tobiasKaminsky should it be merged? should it be backported?

**Comment by tobiasKaminsky** [comment]
:+1: 
fine for me, tested & works

I think, we can merge it, but no need of a backport.
Backport should always be only bugs. Although backports are very easy due to our bot…

**Comment by jancborchardt** [comment]
After looking at it again I had the initial confusion of "But how do I retry a failed upload?" since only the delete action is so present. We should probably have a 3-dot icon which includes both "Retry" and "Cancel" (and cancel using the x icon, not delete as it does not delete a file).

**Comment by tobiasKaminsky** [comment]
@jancborchardt this one is already merged, can you open up a new ticket, so that we can track it better?

**Comment by jancborchardt** [comment]
Sure, done at https://github.com/nextcloud/android/issues/3618 :)
